boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,829 @@
+
+    // Main search loop. We browse the B-Tree which contains directory entries
+    // SI contains the current index entry.
+    NodeCheck:
+        test  word ptr es:[si + 12], 2
+        jnz   NodeCheckLastNode
+
+        mov   cl, byte ptr es:[si + HEX(50)]
+        movzx cx, cl
+        lea   si, [si + HEX(52)]
+        mov   di, offset FreeLdr
+        call  CompareStr
+        lea   si, [si - HEX(52)]
+        test  ax, ax
+        jz    RootIndexFound
+        test  word ptr es:[si + 12], 1
+        jz    ContinueSearch
+
+        test  ax, HEX(f000)
+        jnz   LookupChildNode                  // if result < 0 then explore child node
+
+        ContinueSearch:
+        add   si, word ptr es:[si + 8]
+        jmp   NodeCheck
+
+        RootIndexFound:
+        mov   eax, dword ptr es:[si]           // We found root entry, return with its MFT number
+        mov   edx, dword ptr es:[si + 4]       // Return high part into edx
+        jmp   ExitIndexTree
+    
+        NodeCheckLastNode:
+        test  word ptr es:[si + 12], 1
+        jz    PrintFreeldrError
+
+        LookupChildNode:
+        // Take the right LCN at the end of the index record
+        add   si, word ptr es:[si + 8]
+        mov   ecx, dword ptr es:[si - 8]
+
+        // Read the fixed up LCN
+        mov   bx, word ptr [bp - allocationRunList]
+        mov   ax, word ptr [bp - fileRecordBase]
+        mov   es, ax
+        mov   ax, HEX(9000)
+        mov   fs, ax
+        mov   ax, bx
+        xor   bx, bx
+        call  ReadINDX
+
+        // Go again to the loop but with the child node list
+        mov   ax, HEX(9000)
+        mov   es, ax
+        mov   si, 0
+        add   si, es:[si + HEX(18)] // Go to the first node
+        lea   si, [si + HEX(18)]
+        jmp   NodeCheck
+
+    ExitIndexTree:
+    pop   di
+    pop   si
+    mov   sp, bp
+    pop   bp
+    ret
+
+// [ES:SI] - the source file name
+// [DS:DI] - the destination file name
+// CX - compare length
+CompareStr:
+    push bp
+    mov  bp, sp
+    push bx
+    push si
+    push di
+    movzx cx, cl
+    CmpLoop:
+        mov ax, word ptr es:[si]
+        cmp ax, 'a'
+        jl NoUpper
+        cmp ax, 'z'
+        jg NoUpper
+        sub ax, HEX(20)
+        NoUpper:
+        mov bx, word ptr ds:[di]
+        sub bx, ax
+        test bx, bx
+        jnz  CompareFail
+        add  si, 2
+        add  di, 2
+        dec cx
+        jnz CmpLoop
+    CompareFail:
+    mov ax, bx
+    pop di
+    pop si
+    pop bx
+    mov sp, bp
+    pop bp
+    ret
+
+// Reads a NTFS cluster
+// AX     : cluster number to read
+// ES:BX  : address to read
+ReadCluster:
+    push  edx
+    push  ecx
+    mov   cl, byte ptr [SectsPerCluster]   // Convert sectors to clusters
+    movzx ecx, cx
+    mul   ecx
+    add   eax, dword ptr [HiddenSectors]   // Add offset from the disk beginning
+    call  ReadSectors
+    pop   ecx
+    pop   edx
+    ret
+
+// Reads a MFT entry
+// AX     : MFT inode
```

Comment:
Just make it `eax`, that way you can avoid `movzx eax, ax` stuff (it is already called like that actually)